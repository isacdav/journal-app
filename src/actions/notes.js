import Swal from 'sweetalert2';
import { db } from '../firebase/firebase';
import { fileUpload } from '../helpers/fileHelper';
import { loadNotes } from '../helpers/notesHelper';
import { types } from '../types/types';

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime(),
    };

    const doc = await db.collection(`${uid}/journal/notes`).add(newNote);

    dispatch(activateNote(doc.id, newNote));
    dispatch(addNewNote(doc.id, newNote));
  };
};

export const addNewNote = (id, note) => ({
  type: types.notesAddNew,
  payload: {
    id,
    ...note,
  },
});

export const startLoadingNotes = uid => {
  return async dispatch => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const activateNote = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note,
  },
});

export const setNotes = notes => ({
  type: types.notesLoad,
  payload: notes,
});

export const startSaveNote = note => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    if (!note.url) delete note.url;

    const noteToDb = { ...note };
    delete noteToDb.id;

    await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToDb);

    dispatch(refreshNote(note.id, noteToDb));

    Swal.fire('Saved', note.title, 'success');
  };
};

export const refreshNote = (id, note) => ({
  type: types.notesUpdated,
  payload: {
    id,
    note: {
      id,
      ...note,
    },
  },
});

export const startUploadingFile = file => {
  return async (dispatch, getState) => {
    const { active } = getState().notes;

    Swal.fire({
      title: 'Uploading...',
      text: 'Please wait while the picture is being uploaded.',
      allowOutsideClick: false,
      willOpen: () => {
        Swal.showLoading();
      },
    });

    const fileUrl = await fileUpload(file);
    active.url = fileUrl;

    dispatch(startSaveNote(active));

    Swal.close();
  };
};

export const startDeleting = id => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;
    await db.doc(`${uid}/journal/notes/${id}`).delete();

    dispatch(deleteNote(id));
  };
};

export const deleteNote = id => ({
  type: types.notesDelete,
  payload: id,
});

export const cleanNotesLogout = () => ({
  type: types.notesLogoutCleaning,
});
