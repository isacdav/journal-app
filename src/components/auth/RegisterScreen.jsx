import React from 'react';
import { Link } from 'react-router-dom';

const RegisterScreen = () => {
  return (
    <>
      <h2 className='auth__title mb-20'>Hello</h2>
      <form>
        <input
          className='auth__input'
          type='text'
          placeholder='Name'
          name='name'
          autoComplete='off'
        />
        <input
          className='auth__input'
          type='email'
          placeholder='Email'
          name='email'
          autoComplete='off'
        />
        <input
          className='auth__input'
          type='password'
          placeholder='Password'
          name='password'
        />
        <input
          className='auth__input'
          type='password'
          placeholder='Confirm password'
          name='password2'
        />
        <button className='btn btn-primary btn-block mb-5' type='submit'>
          Log In
        </button>
        <small>
          <Link className='link' to='/auth/login'>
            Already registered?
          </Link>
        </small>
      </form>
    </>
  );
};

export default RegisterScreen;
