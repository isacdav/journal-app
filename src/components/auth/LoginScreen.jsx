import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { startGoogleLogin, startLoginEmail } from '../../actions/auth';
import { removeError, setError } from '../../actions/ui';
import useForm from '../../hooks/useForm';

const LoginScreen = () => {
  const dispatch = useDispatch();
  const { msgError, loading } = useSelector(state => state.ui);

  const [formValues, handleInputChange] = useForm({
    email: '',
    password: '',
  });

  const { email, password } = formValues;

  const handleSubmit = e => {
    e.preventDefault();

    if (!isFormValid()) return;

    dispatch(startLoginEmail(email, password));
  };

  const isFormValid = () => {
    if (email.trim().length <= 0 || password.trim().length <= 0) {
      dispatch(setError('Please enter Email and Password'));
      return false;
    }

    dispatch(removeError());
    return true;
  };

  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  };

  return (
    <>
      <h2 className='auth__title mb-20'>Hello</h2>
      <form
        onSubmit={handleSubmit}
        className='animate__animated animate__fadeIn animate__faster'>
        {msgError && <div className='auth__alert-error'>{msgError}</div>}
        <input
          className='auth__input'
          id='input_email'
          type='email'
          placeholder='Email'
          name='email'
          autoComplete='off'
          value={email}
          onChange={handleInputChange}
        />
        <input
          className='auth__input'
          id='input_pass'
          type='password'
          placeholder='Password'
          name='password'
          value={password}
          onChange={handleInputChange}
        />
        <button
          className='btn btn-primary btn-block mb-5'
          type='submit'
          disabled={loading}>
          Log In
        </button>
        <small>
          <Link className='link' to='/auth/register'>
            New account
          </Link>
        </small>

        <div className='auth__social-net mt-20'>
          <div className='google-btn' onClick={handleGoogleLogin}>
            <div className='google-icon-wrapper'>
              <img
                className='google-icon'
                src='https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg'
                alt='google button'
              />
            </div>
            <p className='btn-text'>
              <b>Sign in with Google</b>
            </p>
          </div>
        </div>
      </form>
    </>
  );
};

export default LoginScreen;
