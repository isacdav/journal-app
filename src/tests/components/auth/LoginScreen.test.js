import '@testing-library/jest-dom';
import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import LoginScreen from '../../../components/auth/LoginScreen';
import { MemoryRouter } from 'react-router-dom';
import { startGoogleLogin, startLoginEmail } from '../../../actions/auth';

jest.mock('../../../actions/auth', () => ({
  startGoogleLogin: jest.fn(),
  startLoginEmail: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
  auth: {},
  ui: {
    loading: false,
    msgError: null,
  },
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <LoginScreen />
    </MemoryRouter>
  </Provider>
);

describe('<LoginScreen> Tests', () => {
  beforeEach(() => {
    store.clearActions();
    jest.clearAllMocks();
  });

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should fire startGoogleLogin', () => {
    wrapper.find('.google-btn').prop('onClick')();

    expect(startGoogleLogin).toHaveBeenCalled();
  });

  test('should fire startLoginEmail', () => {
    wrapper.find('#input_email').simulate('change', {
      target: { name: 'email', value: 'test@mail.com' },
    });
    wrapper
      .find('#input_pass')
      .simulate('change', { target: { name: 'password', value: '12345' } });
    wrapper.find('form').prop('onSubmit')({ preventDefault() {} });

    expect(startLoginEmail).toHaveBeenCalledWith('test@mail.com', '12345');
  });
});
