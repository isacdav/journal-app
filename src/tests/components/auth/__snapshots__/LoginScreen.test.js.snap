// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginScreen> Tests should match snapshot 1`] = `
Array [
  <h2
    className="auth__title mb-20"
  >
    Hello
  </h2>,
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      id="input_email"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="email"
      value=""
    />
    <input
      className="auth__input"
      id="input_pass"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block mb-5"
      disabled={false}
      type="submit"
    >
      Log In
    </button>
    <small>
      <a
        className="link"
        href="/auth/register"
        onClick={[Function]}
      >
        New account
      </a>
    </small>
    <div
      className="auth__social-net mt-20"
    >
      <div
        className="google-btn"
        onClick={[Function]}
      >
        <div
          className="google-icon-wrapper"
        >
          <img
            alt="google button"
            className="google-icon"
            src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
          />
        </div>
        <p
          className="btn-text"
        >
          <b>
            Sign in with Google
          </b>
        </p>
      </div>
    </div>
  </form>,
]
`;
